File: LeetCode/java/LC0059_Spiral_Matrix_II/SpiralMatrixII.java
package LC0059_Spiral_Matrix_II;

import java.util.Arrays;

/**
 * <a href="https://leetcode.cn/problems/spiral-matrix-ii/">Spiral Matrix II</a>
 * 数组；矩阵；模拟
 */
public class SpiralMatrixII {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(Arrays.deepToString(solution.generateMatrix(3)));
        System.out.println(Arrays.deepToString(solution.generateMatrix(1)));
    }
}

class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        int cur = 1;

        int i = 0, j = 0;
        while (cur <= n * n) {
            while (j < n && res[i][j] == 0) {
                res[i][j++] = cur++;
            }
            j--;
            i++;
            while (i < n && res[i][j] == 0) {
                res[i++][j] = cur++;
            }
            i--;
            j--;
            while (j >= 0 && res[i][j] == 0) {
                res[i][j--] = cur++;
            }
            j++;
            i--;
            while (i >= 0 && res[i][j] == 0) {
                res[i--][j] = cur++;
            }
            i++;
            j++;
        }
        return res;
    }
}

File: LeetCode/go/LC0059_Spiral_Matrix_II/spiral_matrix_ii.go
package spiral_matrix_ii

func generateMatrix(n int) [][]int {
	res := make([][]int, n)
	for i := 0; i < n; i++ {
		res[i] = make([]int, n)
	}
	cur := 1

	i, j := 0, 0
	for cur <= n*n {
		for j < n && res[i][j] == 0 {
			res[i][j] = cur
			cur++
			j++
		}
		j--
		i++
		for i < n && res[i][j] == 0 {
			res[i][j] = cur
			cur++
			i++
		}
		i--
		j--
		for j >= 0 && res[i][j] == 0 {
			res[i][j] = cur
			cur++
			j--
		}
		j++
		i--
		for i >= 0 && res[i][j] == 0 {
			res[i][j] = cur
			cur++
			i--
		}
		i++
		j++
	}
	return res
}
